internal/zinc-core/src/main/scala/sbt/internal/inc/InvalidationProfiler.scala

```diff
@@ -0,0 +1,267 @@
+package sbt.internal.inc
+
+import java.io.File
+
+import xsbti.UseScope
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+/**
+ * Defines a profiler interface that translates to the profiling protobuf schema.
+ *
+ * The protobuf schema has been mildly inspired from pprof:
+ * https://github.com/google/pprof/blob/master/proto/profile.proto
+ *
+ * A profiler interface should only be used by project, and not globally, as
+ * this class is not thread safe.
+ */
+abstract class InvalidationProfiler {
+  def profileRun: RunProfiler
+  def registerRun(run: zprof.ZincRun): Unit
+}
+
+object InvalidationProfiler {
+  final val empty: InvalidationProfiler = new InvalidationProfiler {
+    override def profileRun: RunProfiler = RunProfiler.empty
+    override def registerRun(run: zprof.ZincRun): Unit = ()
+  }
+}
+
+class ZincInvalidationProfiler extends InvalidationProfiler {
+  private final var lastKnownIndex: Long = -1L
+  private final val stringTable1: ArrayBuffer[String] = new ArrayBuffer[String](1000)
+  private final val stringTable2: ArrayBuffer[String] = new ArrayBuffer[String](10)
+  private final val stringTableIndices: mutable.HashMap[String, Long] =
+    new mutable.HashMap[String, Long]
+
+  def profileRun: RunProfiler = new ZincProfilerImplementation
+
+  private final var runs: List[zprof.ZincRun] = Nil
+  def registerRun(run: zprof.ZincRun): Unit = {
+    runs = run :: runs
+    ()
+  }
+
+  /**
+   * Returns an immutable zprof profile that can be serialized.
+   *
+   * It is recommended to only perform this operation when we are
+   * going to persist the profiled protobuf data to disk. Do not
+   * call this function after every compiler iteration as the aggregation
+   * of the symbol tables may be expensive, it's recommended to
+   * persist this file periodically.
+   *
+   * @return An immutable zprof profile that can be persisted via protobuf.
+   */
+  def toProfile: zprof.Profile = zprof.Profile(
```

Comment:
This is *the* most important method in the whole implementation, right?
It creates a serializable format for the profile data  gathered so far. Why isn't this part
of the public API?

Reply:
Because this exposes internal details of the algorithm, and I only want people that really know what they're doing using it :wink:

Reply:
okay :smile: